lib/queryContextCache.js

```diff
@@ -0,0 +1,391 @@
+    //     },
+    //     {
+    //       "id": 2,
+    //       "readtimestamp": 123456789,
+    //       "priority": 2,
+    //       "context": "base64 encoded context"
+    //     }
+    //   ]
+
+    const entries = data.entries;
+    if (entries !== null && Array.isArray(entries)) {
+      for (const entryNode of entries) {
+        const entry = this.deserializeQueryContextElement(entryNode);
+        if (entry != null) {
+          this.merge(entry.getId(), entry.getReadTimestamp(), entry.getPriority(), entry.getContext());
+        } else {
+          Logger.getInstance().warn(
+            'deserializeQueryContextJson: deserializeQueryContextElement meets mismatch field type. Clear the QueryContextCache.');
+          this.clearCache();
+          return;
+        }
+      }
+      // after merging all entries, sync the internal priority map to priority map. Because of
+      // priority swicth from GS side,
+      // there could be priority key conflict if we directly operating on the priorityMap during
+      // a round of merge.
+      this.syncPriorityMap();
+    }
+  } catch (e) {
+    Logger.getInstance().debug(`deserializeQueryContextJson: Exception = ${e.getMessage}`, );
+
+    // Not rethrowing. clear the cache as incomplete merge can lead to unexpected behavior.
+    this.clearCache();
+  }
+
+  // After merging all entries, truncate to capacity
+  this.checkCacheCapacity();
+
+  // Log existing cache entries
+  this.logCacheEntries();
+}; // Synchronized
+
+QueryContextCache.prototype.deserializeQueryContextElement = function (node) {
+  const entry = new QueryContextElement();
+  const idNode = node.id;
+  if (typeof idNode === 'number') {
+    entry.setId(idNode);
+  } else {
+    Logger.getInstance().warn('deserializeQueryContextElement: `id` field is not Number type');
+    return null;
+  }
+
+  const timestampNode = node.timestamp;
+  if (typeof timestampNode === 'number') {
+    entry.setReadTimestamp(timestampNode);
+  } else {
+    Logger.getInstance().warn('deserializeQueryContextElement: `timestamp` field is not Long type');
+    return null;
+  }
+
+  const priorityNode = node.priority;
+  if (typeof priorityNode === 'number') {
+    entry.setPriority(priorityNode);
+  } else {
+    Logger.getInstance().warn('deserializeQueryContextElement: `priority` field is not Long type');
+    return null;
+  }
+
+  const contextNode = node.context;
+  if (typeof contextNode === 'string') {
+
+    entry.setContext(contextNode);
+
+  } else if (contextNode === null || contextNode === undefined) {
+    
+    entry.setContext(null);
+    Logger.getInstance().debug('deserializeQueryContextElement `context` field is empty');
+  } else {
+    Logger.getInstance().warn('deserializeQueryContextElement: `context` field is not String type');
+    return null;
+  }
+  return entry;
+};
+
+QueryContextCache.prototype.logCacheEntries = function () {
+  if(Logger.getInstance().getLevel() === 3){
```

Comment:
what is log level 3? let's use names if we really need it?

Reply:
maybe let's use instead `getLevelTag() === LOG_LEVEL_TAGS.DEBUG`

but it won't work for TRACE level (which is 4)?

maybe we should use instead `>=3`?

Reply:
I will use it. Actually, I also feel weird, but all other drivers also capture this log when the log level is only the debug level.

Reply:
I added getLevelTag function in the browser.js because getLeveLTag is existing in Core, and replace LogLevel()===3 with LogLevelTag() === 'DEBUG'

Reply:
I don't think it's right - when I increase log level to TRACE I don't want to receive less logs

maybe we should create a method `Logger.getInstance().logIfLevelAtLeast('DEBUG', () => "Log message")` or `Logger.getInstance().logLevelAtLeast('DEBUG')` (returning boolean)?

Reply:
Actually, we do not need to create the method you mentioned above if we want to print out the debug level log at the higher level. Logger.getInstance().debug() function will do the same thing. As a result, please let me know if we are okay to print this log in the log level higher than the "Debug" Level. According to the comments, it said it is a debugging purposes only.

Here is
JDBC: https://github.com/snowflakedb/snowflake-jdbc/pull/1194/files#diff-9074198cac88258f7cbb81d705f00b8441e0c6b18c9378ef8e68ea7c29a074eb
.Net: https://github.com/snowflakedb/snowflake-connector-net/pull/757/files#diff-fd16d3594fdc6c87196c0471ed9507e1d198b47997b1761534f9eb4eee2a6e85

Reply:
Still - on TRACE level we should see the logs from ERROR, WARN, INFO and DEBUG
In JDBC `isDebugEnabled` should work also with `TRACE` level
This function is used because there are complex calculations to create the debug log - but definitely when on TRACE we also want to see the log message here

Reply:
Sure, Then I will change this. Thanks.